src/graphics/uniform-buffer-format.js

```diff
@@ -66,14 +61,37 @@ class UniformFormat {
     /** @type {ScopeId} */
     scopeId;
 
-    // TODO: add count for arrays
+    /**
+     * Count of elements for arrays, otherwise 1.
+     *
+     * @type {number}
+     */
+    count;
 
-    constructor(name, type) {
+    constructor(name, type, count = 1) {
         this.name = name;
         this.type = type;
-        this.byteSize = uniformTypeToByteSize[type];
+
+        this.count = count;
+        Debug.assert(count === 1, `Uniform arrays are not currently supported - uniform ${name}`);
+
+        const elementSize = uniformTypeToNumElements[type];
+        Debug.assert(elementSize, `Unhandled uniform format ${type} used for ${name}`);
+
+        this.byteSize = count * elementSize * 4;
         Debug.assert(this.byteSize, `Unknown byte size for uniform format ${type} used for ${name}`);
     }
+
+    // std140 rules: https://registry.khronos.org/OpenGL/specs/gl/glspec45.core.pdf#page=159
+    calculateOffset(offset) {
+
+        // Note: vec3 has the same alignment as vec4
+        const alignment = this.byteSize <= 8 ? this.byteSize : 16;
```

Comment:
This is correct for single values, but it's incomplete for arrays and structs where it has to be run recursively. 

Perhaps note this down so we don't forget it, as alignment issues are rather hard to detect.

Reply:
adding comment